core/unix/signal.c

```diff
@@ -8329,6 +8328,39 @@ sig_detach(dcontext_t *dcontext, sigframe_rt_t *frame, KSYNCH_TYPE *detached)
     ASSERT_NOT_REACHED();
 }
 
+static bool
+is_signal_for_us(kernel_siginfo_t *siginfo, int sig_expecting)
+{
+    /* Distinguish a nudge/suspend from an app signal.
+     * On Linux with SYS_rt_tgsigqueueinfo support, it is much easier, since an
+     * app using libc sigqueue()
+     * will never have its signal mistaken as libc does not expose the kernel_siginfo_t
+     * and always passes 0 for si_errno, so we're only worried beyond our
+     * si_code check about an app using a raw syscall that is deliberately
+     * trying to fool us.
+     * While there is a lot of padding space in kernel_siginfo_t, the kernel doesn't
+     * copy it through on SYS_rt_sigqueueinfo so we don't have room for any
+     * dedicated magic numbers.  The client id could function as a magic
+     * number for client nudges, but I don't think we want to kill the app
+     * if an external nudger types the client id wrong.
+     */
+    LOG(THREAD_GET, LOG_ASYNCH, 2, "%s T%d: sig=%d code=%d errno=%d\n", __FUNCTION__,
+        get_sys_thread_id(), siginfo->si_signo, siginfo->si_code, siginfo->si_errno);
+    if (siginfo->si_signo != sig_expecting)
+        return false;
+    if (is_sigqueue_supported()) {
+        if (siginfo->si_code != SI_QUEUE || siginfo->si_errno == 0)
+            return false;
+    } else {
+        /* Distinguish from a synchronous app signal by looking for
+         * si_code where <=0 means user-sent.
+         */
+        if (siginfo->si_code > 0)
```

Comment:
So our signal will have > 0? Do we set it ourselves? if not, how is it set?

Reply:
The kernel sets this field.